src/Kestrel.Transport.Sockets/Internal/SocketConnection.cs

```diff
@@ -255,5 +267,36 @@ internal SocketConnection(Socket socket, MemoryPool<byte> memoryPool, PipeSchedu
                 }
             }
         }
+
+        private void Shutdown()
+        {
+            lock (_shutdownLock)
+            {
+                if (!_aborted)
+                {
+                    // Make sure to close the connection only after the _aborted flag is set.
+                    // Without this, the RequestsCanBeAbortedMidRead test will sometimes fail when
+                    // a BadHttpRequestException is thrown instead of a TaskCanceledException.
+                    _aborted = true;
+                    _trace.ConnectionWriteFin(ConnectionId);
+
+                    // Try to gracefully close the socket even for aborts to match libuv behavior.
+                    _socket.Shutdown(SocketShutdown.Both);
```

Comment:
codifyd:
```
            Socket serverSocket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
            Socket clientSocket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
            serverSocket.Bind(new IPEndPoint(IPAddress.Loopback, 0));
            serverSocket.Listen(10);
            clientSocket.Connect(serverSocket.LocalEndPoint);
            Socket acceptedSocket = serverSocket.Accept();
            acceptedSocket.Dispose();
            byte[] buffer = new byte[10];
            int received = clientSocket.Receive(buffer);
            System.Console.WriteLine(received);
```
This prints out `0`.
If `acceptedSocket.Dispose();` is replaced by `acceptedSocket.Close(0);` then it throws `SocketException: Connection reset by peer`.

Reply:
@tmds I tried Dispose instead of Shutdown originally, but it causes issues where the client sees unintended connection resets on Windows. This is one of hundreds of failures that happens in the Sockets functional tests if you replace this `_socket.Shutdown(SocketShutdown.Both)` call with `_socket.Dispose()`: 

```
    Http10KeepAlive(listenOptions: http://127.0.0.1:0) [FAIL]
      System.IO.IOException : Unable to read data from the transport connection: An existing connection was forcibly closed by the remote host.
      ---- System.Net.Sockets.SocketException : An existing connection was forcibly closed by the remote host
      Stack Trace:
           at System.IO.StreamReader.ReadBufferAsync()
           at System.IO.StreamReader.ReadAsyncInternal(Memory`1 buffer, CancellationToken cancellationToken)
           at Microsoft.AspNetCore.Testing.TaskExtensions.TimeoutAfter[T](Task`1 task, TimeSpan timeout, String filePath, Int32 lineNumber)
        G:\dev\aspnet\KestrelHttpServer-Private\test\shared\TestConnection.cs(131,0): at Microsoft.AspNetCore.Testing.TestConnection.Receive(String[] lines)
        G:\dev\aspnet\KestrelHttpServer-Private\test\shared\TestConnection.cs(162,0): at Microsoft.AspNetCore.Testing.TestConnection.ReceiveForcedEnd(String[] lines)
        G:\dev\aspnet\KestrelHttpServer-Private\test\Kestrel.FunctionalTests\RequestTests.cs(910,0): at Microsoft.AspNetCore.Server.Kestrel.FunctionalTests.RequestTests.Http10KeepAlive(ListenOptions listenOptions)
        --- End of stack trace from previous location where exception was thrown ---
        ----- Inner Stack Trace -----
```

Reply:
I changed SocketConnection.cs (dev branch) to Dispose rather than Shutdown. On Linux, all tests pass. It would be interesting to have a minimal repro that makes Windows reset the connection to know if this is the expected behavior.